### ISSUE TITLE
How to expose fd_set

### ISSUE THREAD

**Issue by TiagoCardoso1983** [ffi/ffi#464]
I would like to turn an array of IO objects into an fd_set pointer I can pass to C functions which require one:

```ruby
readers = IO.select([fd_to_read])
readers_set = mightymophingpowerfdset(readers)
C.functioncallwhichrequiresfdset(readers_set)
```

I haven't found anything in the wiki regarding fd sets, but I'll keep looking.

**Comment by larskanis** [comment]
There is no builtin way to use fd_set with FFI. You must create the `struct fd_set` per `FFI::MemoryPointer.new` (or better derive your own class from `FFI::Struct` ) and fill it with the file descriptors retrieved per `IO#fileno` .

However the `struct fd_set` is defined differently on different platforms. Windows uses an array of 64 file descriptors and Linux uses a bitmask of 1024 bits. Since you can't use `FD_SET` etc. macros in Ruby, you need to implement your own methods depending on the platform.

See https://msdn.microsoft.com/de-de/library/windows/desktop/ms737873(v=vs.85).aspx and http://stackoverflow.com/questions/18952564/understanding-fd-set-in-unix-sys-select-h

**Comment by TiagoCardoso1983** [comment]
Sweet beJesus!!! Thx for the info, I guess it's doable then. Couldn't the FFI library support the fd_set struct as an opt-in feature and therefore saving the platform-compatibility trouble? Some of the ported C libraries that have interest to be used within Ruby handle non-blocking I/O and the most used (that I know of) implementations (EM, celluloid-io) resort to writing the c-extensions themselves. 

In case of a positive answer, I'll publish later what I managed.

**Comment by tduehr** [comment]
I'm not sure this is the realm of FFI. `select` is implemented in Ruby already. Gems handling non-blocking IO have C extensions because using FFI or pure ruby would make the library too slow. There are also other platform options, such as epoll and kqueue, that are much better than select.

**Comment by larskanis** [comment]
@TiagoCardoso1983 I don't think we will integrate special features for `struct fd_set` into FFI directly. However you have several alternative options:

1. Use the pure FFI based encoding of `struct fd_set` described [above](https://github.com/ffi/ffi/issues/464#issuecomment-149867946) . This will run on JRuby as well.

2. You write a new gem as a Ruby C-ext, which does the conversions to/from `struct fd_set` in C. It can make use of FD_SET etc. macros, so it should be more portable across different operating systems. It however will not run on JRuby.

3. You make use of ffi-inline.gem which builds a shared object or DLL file to be loaded by FFI. It can equally make use of FD_SET etc. macros. It should be able to run on JRuby.

**Comment by TiagoCardoso1983** [comment]
thx for the comments. I'll proceed with option 1. Is 3. your ideal solution?

**Comment by tduehr** [comment]
1 and 3 are functionally equivalent with 1 having slightly better testing and support. 2 will get you the best speed but at the cost of portability across Ruby implementations.

**Comment by larskanis** [comment]
I would take option 1 with an implementation for Linux and Windows. OS-X is probably equally to Linux.

However as @tduehr pointed out, this all is only really useful, if you want to interface with a library that explicitly requires the use of `struct fd_set` .

**Comment by TiagoCardoso1983** [comment]
https://github.com/chuckremes/ffi-rzmq-core/blob/master/lib/ffi-rzmq-core/structures.rb#L25-L70 could this be a step in the right direction?

**Comment by tduehr** [comment]
@larskanis FWIW here's OSX:

```
typedef	struct fd_set {
	__int32_t	fds_bits[__DARWIN_howmany(__DARWIN_FD_SETSIZE, __DARWIN_NFDBITS)];
} fd_set;
```

@TiagoCardoso1983 Yes, that seems to be the correct route.
